flink-runtime/src/main/java/org/apache/flink/runtime/operators/shipping/OutputEmitter.java

```diff
@@ -143,16 +148,24 @@ public OutputEmitter(ShipStrategyType strategy, TypeComparator<T> comparator, Pa
 	
 	// --------------------------------------------------------------------------------------------
 
+	private int[] forward() {
+		return this.channels;
+	}
+
 	private int[] robin(int numberOfChannels) {
-		if (this.channels == null || this.channels.length != 1) {
-			this.channels = new int[1];
+		int nextChannel = this.nextChannelToSendTo;
+
+		if (nextChannel >= numberOfChannels) {
+			if (nextChannel == numberOfChannels) {
+				nextChannel = 0;
```

Comment:
The modulo computation is executed in two cases: 1) the initial call to robin when the source parallelism is greater than the receiver parallelism and 2) when the receiver parallelism is dynamically changing.

Unoptimized modulo is much slower than a test for equality but is required in the first case to evenly distribute the first outputs.